webapi/src/main/scala/org/knora/webapi/responders/v2/ontology/Cardinalities.scala

```diff
@@ -250,9 +262,24 @@ object Cardinalities {
       cardinalitiesToDelete: Map[SmartIri, Cardinality.KnoraCardinalityInfo] =
         deleteCardinalitiesFromClassRequest.classInfoContent.toOntologySchema(InternalSchema).directCardinalities
 
-      _ = cardinalitiesToDelete.foreach(p =>
-        isCardinalityDefinedOnClass(cacheData, p._1, p._2, internalClassIri, internalOntologyIri)
-      )
+      isDefinedOnClassList: List[Boolean] <- Future
+        .sequence(cardinalitiesToDelete.map { p =>
+          for {
+            isDefined: Boolean <- isCardinalityDefinedOnClass(
+              cacheData,
+              p._1,
+              p._2,
+              internalClassIri,
+              internalOntologyIri
+            )
+          } yield isDefined
+        }.toList)
+
+      _ = if (isDefinedOnClassList.contains(false)) {
+        throw BadRequestException(
+          "The cardinality is not defined directly on the class and cannot be deleted."
+        )
+      }
```

Comment:
this seems to be pretty much exactly the same as you do above in the `canDelete...` (except throwning an exception instead of returning a boolean), so it would probably be cleaner to extract it into a separate function

Reply:
I agree that is quite a lot of code copied but I actually don't see right now how to simplify it. `isCardinalityDefinedOnClass` has to be called on the Map`cardinalitiesToDelete`. I don't think it helps much to have a separate function that just "loops" through (maps) the list and returns a Future of a list (then, of course, it would make more sense to implement a higher-order function for this kind of action but I'm not sure this makes sense now before the greater refactoring we do anyway). So I leave it for now.